Show full repr with assert a==b and -vv

oscarbenjamin opened the pull request:
Fixes #2256 

Adds a new explanation handler to show reprs of arbitrary objects. The handler shows the full `repr` in test failure output for `assert x==y` when `-vv` is given on the command line.

Here's a test case:
```python
class Nums:
    def __init__(self, nums):
        self.nums = nums
    def __repr__(self):
        return str(self.nums)

def test_f():
    x = list(range(5000))
    y = list(range(5000))
    y[len(y)//2] = 3
    x = Nums(x)
    y = Nums(y)
    assert x == y
```

Running on master this gives:
```terminal
$ pytest test_f.py -vv
==================================================================== test session starts =====================================================================
platform darwin -- Python 3.7.1, pytest-4.0.3.dev19+ge8152207, py-1.7.0, pluggy-0.8.0 -- /Users/enojb/current/sympy/venv/bin/python3
cachedir: .pytest_cache
rootdir: /Users/enojb/current/sympy/pytest, inifile: tox.ini
plugins: xdist-1.24.1, forked-0.2, doctestplus-0.3.0.dev0
collected 1 item                                                                                                                                             

test_f.py::test_f FAILED                                                                                                                               [100%]

========================================================================== FAILURES ==========================================================================
___________________________________________________________________________ test_f ___________________________________________________________________________

    def test_f():
        x = list(range(5000))
        y = list(range(5000))
        y[len(y)//2] = 3
        x = Nums(x)
        y = Nums(y)
>       assert x == y
E       assert [0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31,...4980, 4981, 4982, 4983, 4984, 4985, 4986, 4987, 4988, 4989, 4990, 4991, 4992, 4993, 4994, 4995, 4996, 4997, 4998, 4999] == [0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31,...4980, 4981, 4982, 4983, 4984, 4985, 4986, 4987, 4988, 4989, 4990, 4991, 4992, 4993, 4994, 4995, 4996, 4997, 4998, 4999]

test_f.py:13: AssertionError
================================================================== short test summary info ===================================================================
FAIL test_f.py::test_f
================================================================== 1 failed in 0.09 seconds ==================================================================
```

With this patch there is a short output given by:
```terminal
$ pytest test_f.py -v
==================================================================== test session starts =====================================================================
platform darwin -- Python 3.7.1, pytest-4.0.3.dev19+ge8152207, py-1.7.0, pluggy-0.8.0 -- /Users/enojb/current/sympy/venv/bin/python3
cachedir: .pytest_cache
rootdir: /Users/enojb/current/sympy/pytest, inifile: tox.ini
plugins: xdist-1.24.1, forked-0.2, doctestplus-0.3.0.dev0
collected 1 item                                                                                                                                             

test_f.py::test_f FAILED                                                                                                                               [100%]

========================================================================== FAILURES ==========================================================================
___________________________________________________________________________ test_f ___________________________________________________________________________

    def test_f():
        x = list(range(5000))
        y = list(range(5000))
        y[len(y)//2] = 3
        x = Nums(x)
        y = Nums(y)
>       assert x == y
E       AssertionError: assert [0, 1, 2, 3, ...7, 4998, 4999] == [0, 1, 2, 3, 4...7, 4998, 4999]
E         -[0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31, 32, 33, 34, 35, 36, 37, 38, 39, 40, 41, 42, 43, 44, 45, 46, 47, 48, 49, 50, 51, 52, 53, 54, 55, 56, 57, 58, 59, 60, 61, 62, 63, 64, 65, 66, 67, 68, 69, 70, 71, 72, 73, 74, 75, 76, 77, 78, 79, 80, 81, 82, 83, 84, 85, 86, 87, 88, 89, 90, 91, 92, 93, 94, 95, 96, 97, 98, 99, 100, 101, 102, 103, 104, 105, 106, 107, 108, 109, 110, 111, 112, 113, 114, 115, 116, 117, 118, 119, 120, 121, 122, 123, 124, 125, 126, 127, 128, 129, 130, 131, 132, 133, 134, 135, 136...
E         
E         ...Full output truncated (2 lines hidden), use '-vv' to show

test_f.py:13: AssertionError
================================================================== short test summary info ===================================================================
FAIL test_f.py::test_f
================================================================== 1 failed in 0.10 seconds ==================================================================
```

Running with `-vv` shows the whole `repr` (which is very long so I won't paste it here).

blueyed commented:
Thanks, please add a test for this - (maybe only by adjusting the now failing test(s)?!)

RonnyPfannschmidt commented on src/_pytest/assertion/util.py:
```diff
@@ -227,6 +229,20 @@ def escape_for_readable_diff(binary_text):
+    left_lines = repr(left).splitlines(keepends)
+    right_lines = repr(right).splitlines(keepends)
+
+    explanation = []
```
`unified_diff` imported but not used

blueyed commented:
It might be worth controlling this through some additional setting.

I for example would like to see full diffs always, without turning pytest verbose in general.

oscarbenjamin commented:
What verbosity level should I use?

It's already easy to make arbitrarily long output with `-vv` using strings. I can also make `pytest` hang in `ndiff` easily enough if I want:
```python
import random

def test_f():
    linesa = '\n'.join('foobar' + random.choice('wer') for n in range(10000))
    linesb = '\n'.join('foobar' + random.choice('wer') for n in range(10000))
    assert linesa == linesb
```
(`ndiff` is very slow: [https://bugs.python.org/issue6931](https://bugs.python.org/issue6931)

The behaviour I've gone for here is intended to match what already happens with strings.

nicoddemus commented:
> It might be worth controlling this through some additional setting.

I also have the same feeling. How about `--full-diff`?

nicoddemus commented:
We also need a new CHANGELOG entry

nicoddemus commented:
Adding a new option will require this to target `features`.

oscarbenjamin commented:
I think I've addressed all issues (removed unused import, added changelog, fixed it so no tests fail, added a new test) apart from the question of the command line option to use.

I don't understand why there should be a separate option unless it is intended to apply to all the cases. The behaviour of this patch matches what already happens for list, set, string etc.

If you try this you can see the effect of `-v` and `-vv` for all the other types:
```python

def test_f():
    class Nums:
        def __init__(self, nums):
            self.nums = nums
        def __repr__(self):
            return str(self.nums)

    x = list(range(500))
    y = list(range(500))
    y[len(y)//2] = 3
    x = Nums(x)
    y = Nums(y)
    assert x == y

def test_g():
    x = list(range(500))
    y = list(range(500))
    y[len(y)//2] = -1
    assert x == y

def test_h():
    x = set(range(500))
    y = set(range(500))
    y.remove(200)
    assert x == y

def test_i():
    lines = ['qwerty'*5 for n in range(20)]
    x = '\n'.join(lines)
    lines[10] += 'zxc'
    y = '\n'.join(lines)
    assert x == y

def test_j():
    x = list(range(500))
    y = list(range(500))
    y[len(y)//2] = 3
    x = str(x)
    y = str(y)
    assert x == y
```

You can see that `verbosity=0` (no `-v` arg) gives one short line of output.  Passing `-v` (`verbosity=1`) gives a few longer lines. Passing `-vv` (`verbosity=2`) gives the full output. This is true in all cases except for `test_f` which on current master will not give full output but will with this patch.

If there is to be a separate command line option why should it only apply to this case and not the others?

Likewise I don't see how this case can cause a *new* DOS issue given that it matches the other cases.

nicoddemus commented:
Well, as a first interaction we can keep it as `-vv` then. If we decide to introduce a new option, we can do it later and apply it for all cases, as you mention.

nicoddemus commented on src/_pytest/assertion/util.py:
```diff
@@ -236,6 +238,18 @@ def escape_for_readable_diff(binary_text):
     return explanation
 
 
+def _compare_eq_verbose(left, right, verbose=False):
```
`verbose` is not used?

blueyed commented on src/_pytest/assertion/util.py:
```diff
@@ -151,6 +151,8 @@ def isiterable(obj):
                     type_fn = (isdatacls, isattrs)
                     explanation = _compare_eq_cls(left, right, verbose, type_fn)
+                elif verbose:
+                    explanation = _compare_eq_verbose(left, right)
```
This uses it with `-v` already, no?

oscarbenjamin replied:
Yes. That's also what happens in the case of e.g. strings: [https://github.com/pytest-dev/pytest/blob/master/src/_pytest/assertion/util.py#L176](https://github.com/pytest-dev/pytest/blob/master/src/_pytest/assertion/util.py#L176)

The filtering based on -vv happens down the line somewhere else (in all cases).

So with -v this function will generate full output but pytest won't ultimately show the whole of it unless -vv is given.

oscarbenjamin replied:
The test is showing the output of `call_equal` which calls these functions directly. Shortening (without -vv) takes place here I think: [https://github.com/pytest-dev/pytest/blob/master/src/_pytest/assertion/truncate.py](https://github.com/pytest-dev/pytest/blob/master/src/_pytest/assertion/truncate.py)

That looks like a deliberate design and seems reasonable to me so I'm not sure what needs fixing.

nicoddemus replied:
I think it makes sense that truncation happens later down the line, otherwise every implementation would need to implement or call truncation (IIUC).

nicoddemus replied:
But yeah, I was also confused for this same reason, I had to look at the rest of the code to understand what was going on.

blueyed replied:
@nicoddemus 
I think we should create a follow-up issue/PR for this.
Have you looked closer at it already?

nicoddemus commented:
Thanks again @oscarbenjamin for the PR!

nicoddemus commented:
Thanks again @oscarbenjamin for the PR!

oscarbenjamin commented:
Thanks all!

blueyed commented:
For reference: this caused a regression (https://github.com/pytest-dev/pytest/issues/5192, https://github.com/pytest-dev/pytest/issues/5932, ), addressed in https://github.com/pytest-dev/pytest/pull/5933.
